builtin/time: support CURRENT_TIME(), CURTIME()

yanzhe-chen opened the pull request:
This is for issue [#236](https://github.com/pingcap/tidb/issues/236). The change adds two time functions: `CURRENT_TIME` and `CURTIME`.

ngaut commented:
Thanks @yanzhe-chen 
Please add some tests.

yanzhe-chen commented:
@ngaut I am not sure how to test these two functions, since its result is real-time decided and accepts no inputs. Could you give me some hints about how to do this?

ngaut commented:
For example, these code are just thoughts:
min := time.Now()
sqlNow := select CURTIME
assert(min < sqlNow < time.Now())

yanzhe-chen commented:
@ngaut Nice! I'll try that and update this request when done.

siddontang commented on mysql/time.go:
```diff
@@ -34,8 +34,9 @@ var (
-	DateFormat = "2006-01-02"
-	TimeFormat = "2006-01-02 15:04:05"
+	DateFormat    = "2006-01-02"
+	CurtimeFormat = "15:04:05"
```
CurTimeFormat or CurrentTimeFormat may be better.

yanzhe-chen replied:
@siddontang Fine, I will make that change.

siddontang commented on expression/builtin/time.go:
```diff
@@ -309,6 +309,13 @@ func builtinCurrentDate(args []interface{}, ctx map[interface{}]interface{}) (in
+func builtinCurrentTime(args []interface{}, ctx map[interface{}]interface{}) (interface{}, error) {
+	return mysql.Time{
+		Time: time.Now(),
+		Type: mysql.TypeDuration, Fsp: 0}, nil
```
I think the type is mysql.TypeTimestamp?

yanzhe-chen replied:
@siddontang `mysql.TypeTimestamp` is for time format: `YYYY-MM-DD HH:MM:SS` and it already has corresponding time functions (`CURRENT_TIMESTAMP()` and `NOW()`).

Actually what I really want to use here is `mysql.TypeTime`, but when I check the code in `mysql/type.go`, the `TypeTime` is renamed to `TypeDuration` and the comment says it has to resolve the conflict with Go type time.

I know it is a bit of confusing but finally I have to adhere to the existing code.

If you have some other workarounds, please let me know.

yanzhe-chen commented:
@ngaut @siddontang I have added the missing test and changed the variable name as suggested. Please review the request again.

ngaut commented:
Good, Please add some unit tests in parser/parser_test.go TestBuiltin.
https://github.com/pingcap/tidb/blob/master/parser/parser_test.go#L329

siddontang commented on expression/builtin/time.go:
```diff
@@ -309,6 +309,26 @@ func builtinCurrentDate(args []interface{}, ctx map[interface{}]interface{}) (in
 }
 
+// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_curtime
+func builtinCurrentTime(args []interface{}, ctx map[interface{}]interface{}) (interface{}, error) {
```
Please use mysql.Duration instead of mysql.Time, or first you can use the mysql.Time to handle the argument, then convert it to mysql.Duration. 

mysql.Time is not for duration, only for datetime, timestamp and date.

coocood commented on expression/builtin/time_test.go:
```diff
@@ -275,3 +275,12 @@ func (s *testBuiltinSuite) TestCurrentDate(c *C) {
+	v, err := builtinCurrentTime(nil, nil)
+	c.Assert(err, IsNil)
+	n, ok := v.(mysql.Time)
+	c.Assert(ok, IsTrue)
```
Please add this assertion 
`c.Assert(n.String(), HasLen, 8)`
and make sure it passes.

shenli commented on parser/scanner.l:
```diff
@@ -692,6 +694,10 @@ year_month		{y}{e}{a}{r}_{m}{o}{n}{t}{h}
 			return curDate
 {current_date}		lval.item = string(l.val)
 			return currentDate
+{curtime}		lval.item = string(l.val)
```
curtime is not MySQL keyword and can be used as identifier. So it should be added into NotKeywordToken rule in parser.y. Those tokens can be used as Identifier.
https://github.com/pingcap/tidb/blob/master/parser/parser.y#L1685

current_time is a reserved keyword. So it can not be used as identifier. It is ok now.

yanzhe-chen commented:
@ngaut I have added the corresponding tests in parser.

By the way, I forgot the the **fsp** arguments as stated in the MySQL doc so I added them as well.

@coocood I used `mysql.Duration` instead and added string length checks in test.

@shenli Thanks for reminding and I am trying to figure out how to fix that.

Please review it again and thanks for your patient help.

coocood commented on parser/parser.y:
```diff
@@ -2137,6 +2139,22 @@ FunctionCallNonKeyword:
 	{
 		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1.(string))}
 	}
+|	"CURTIME" FuncDatetimePrec
```
As "CURTIME" is synonym of "CURRENT_TIME", we can implement one of them, and parse the other one to the same type of token in `scanner.l`, like this PR:
https://github.com/pingcap/tidb/pull/728

yanzhe-chen commented:
@coocood Thanks for your hint and I have made that change.

BTW I have found some other synonyms in code that are not handled in this way (e.g. `curdate` and `current_date`). Shall we create a new issue to fix these inconsistent implementations?

ngaut commented:
Good catch! @yanzhe-chen 
Please feel free to fire issues.

coocood commented:
LGTM
@yanzhe-chen Good job!

siddontang commented on expression/builtin/builtin.go:
```diff
@@ -69,7 +69,9 @@ var Funcs = map[string]Func{
 	"current_date":      {builtinCurrentDate, 0, 0, false, false},
+	"current_time":      {builtinCurrentTime, 0, 0, false, false},
 	"current_timestamp": {builtinNow, 0, 1, false, false},
+	"curtime":           {builtinCurrentTime, 0, 0, false, false},
```
the max arg should be 1. 0 means no argument.

yanzhe-chen commented:
@siddontang Thanks for catching that! Fix it.

yanzhe-chen commented:
@siddontang https://github.com/pingcap/tidb/blob/master/expression/builtin/builtin.go#L73

yanzhe-chen commented on expression/builtin/builtin.go:
```diff
@@ -69,7 +69,9 @@ var Funcs = map[string]Func{
+	"current_time":      {builtinCurrentTime, 0, 1, false, false},
 	"current_timestamp": {builtinNow, 0, 1, false, false},
+	"curtime":           {builtinCurrentTime, 0, 1, false, false},
 	"date":              {builtinDate, 8, 8, true, false},
```
@siddontang Why `builtinDate`'s `minArgs` and `maxArgs` are both set to `8`? `DATE` should accept exactly one argument in the MySQL doc (`DATE(expr)`).
BTW, previously when I set `current_time` and `curtime`'s `maxArgs` to `1`, it can still pass all the tests. This makes me doubt whether this kind of checks really work.
Could you help me explain these? Thanks!

ngaut commented:
Good catch, It's bug which describe here #758

yanzhe-chen commented:
@ngaut Nice, and I tried to fix this issue and fired a new pull request [here](https://github.com/pingcap/tidb/pull/759). Please review it.

yanzhe-chen commented:
@siddontang @ngaut @shenli PTAL if there is anything to improve or we can merge and close this pull.

ngaut commented:
LGTM, thanks @yanzhe-chen
